### PR TITLE
Use HomePage.go()  Instead of HeaderPage.clickOnHomeLink()

### PR DESCRIPTION
**Pull request by sherrif10**
Follow up https://issues.openmrs.org/browse/RATEST-74
 cc @kdaud @dkayiwa @ibacher 
What changed in this pull request
i would not say this would be the fix to the bamboo logs directly due to the failures https://ci.openmrs.org/browse/REFAPP-UI-428 , however this should fix the ci to green without causing any issues
i deleted the functionality in tear down part of extending the automation to headerPage which is resulting into the failures on bamboo by the visibility of an element. One easiest and clean way is to add the delete function in tear down part, this could be the right and easiest way to fix this without harming anything and leaving all the functionalities working as expected

### CODE REVIEW

**Comment by sherrif10** [comment]
cc @kdaud  Looking forward for your review and suggestions.

**Comment by kdaud** [comment]
Thanks @sherrif10 for following up on FormTest.

Here is my thought!

`teardown()` function may not need to be refactored and thoughtfully might not be the solution to the problem.
But rather as I mentioned in the first paragraph [here ](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/523#issuecomment-894965038) the solution is to address the method at the code line [29](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/887b24e27574a50807f1c7c5db4319e8bcad756a/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L115). Certainly the noise is coming from the element look up the function is consuming which is present in the header page [here](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/887b24e27574a50807f1c7c5db4319e8bcad756a/ui-tests/src/test/java/org/openmrs/reference/page/HeaderPage.java#L29). 

In reference to your comment [here ](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/523#issuecomment-895223015) it simply implies that the logout element look up was not solved and that's what is causing FormTest to fail.

I would think looking into that element look up should be the best solution.

Some thought for you @sherrif10 . Have seen you on this commit tagging geeks however, what do you have to say about Daniel's comment [here ](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/522#issuecomment-895443615)?

**Comment by sherrif10** [comment]
> Some thought for you @sherrif10 . Have seen you on this commit tagging geeks however, what do you have to say about Daniel's comment [here ](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/522#issuecomment-895443615)?

All is well. Thanks

**Comment by sherrif10** [comment]
> `teardown()` function may not need to be refactored and thoughtfully might not be the solution to the problem.

Am very sure this will resolve the issue, Like you have commented on[ line 115](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/887b24e27574a50807f1c7c5db4319e8bcad756a/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L115), this is where the problem is , As you saw this commit [here ](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/512), it was to fix the same issue and hence the issue is still the same, i believe this is because the return url of header.Page is simply because its returning null here https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/HeaderPage.java#L33 . As my investigations this morning, HeaderPage is less being used, some of the tests were refactored  not to use headerPage for the moment.

What i did to refactor HeaderPage was once recommended by @ibacher  some times ago, but besides that, this would be the easy was to handle this kind of automation, i strongly believe this will make the bamboo green.

**Comment by kdaud** [comment]
> What i did to refactor HeaderPage was once recommended by @ibacher some times ago, but besides that, this would be the easy was to handle this kind of automation, i strongly believe this will make the bamboo green once merged

In software development, devs suggest ideas and those ideas `may/ may-not` be the ultimate solution but rather as a problem is looked at from different angles then ultimately a better solution is derived, and this is what makes open source development interesting and amazing!!

My opinion to the solution still stands at the third paragraph I mentioned [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/524#issuecomment-895812977).

**Comment by sherrif10** [comment]
> My opinion to the solution still stands at the third paragraph I mentioned [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/524#issuecomment-895812977).

Thanks @kdaud , so is that your final suggestion. i think this is very clear to have fixed the bamboo and hence making things more simpler and test more readable

**Comment by kdaud** [comment]
> i think this is very clear to have fixed the bamboo and hence making things more simpler and test more readable.

@sherrif10 you may need to revisit my recommendation [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/524#issuecomment-895812977) otherwise I am very open to hearing from other reviewers too.

**Comment by sherrif10** [comment]
Good Thanks @kdaud  for the review.

**Comment by sherrif10** [comment]
Waiting on final reviews since @kdaud seems to be open for more reviews

File: ui-tests/src/test/java/org/openmrs/reference/FormTest.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/FormTest.java:100]: Yes sure , its an actual test that deletes the form which i thought would work in tear down part

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/FormTest.java:100]: Let me correct it

**Comment by kdaud** [comment]
Thoughts!

Looking at the old test workflow, we need the `teardown()` function once the test logic has been performed. Suggestively we rather get rid of headPage.logout() within the `teardown() `so that the test execution comes to the end when its at homePage

We can have the tearDown() function as:

@After
	public void tearDown() throws Exception {
		if (headerPage != null) {
			headerPage.clickOnHomeIcon();
		}
	}

We can create a ticket to follow up on the of fix [logout](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/887b24e27574a50807f1c7c5db4319e8bcad756a/ui-tests/src/test/java/org/openmrs/reference/page/HeaderPage.java#L29) element look-up!

**Comment by sherrif10** [comment]
> `@After`
> `public void tearDown() throws Exception` {
> if (headerPage != null) {
> headerPage.clickOnHomeIcon();
> `}`
> }

This is also worth it .i have no problem with its implementation, the idea behind teardown is two things,One is to have a  functionality that is meaningfull

**Comment by sherrif10** [comment]
i also think we definately need to get rid of headerPages, why would we always catch NullpointerExceptions, Most of tests that uses HeaderPages need to handle NullPointerExceptions which i think its not a good practice.

**Comment by kdaud** [comment]
Thought!

We can get rid of redundant code lines to have consistence within our codebase
 eg Line [20 ](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L20) [27](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L27) [29](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L29) [31](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L31) [33](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L33) [35](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L35) [37](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L37) [39](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L39) [41](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L41) [43](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L43) [74](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L74) [88](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L88) [100](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L100)

**Comment by sherrif10** [comment]
> We can get rid of redundant code lines to have consistence within(throughout) our codebase
> eg Line [20 ](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L20) [27](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L27) [29](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L29) [31](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L31) [33](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L33) [35](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L35) [37](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L37) [39](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L39) [41](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L41) [43](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L43) [74](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L74) [88](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L88) [100](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7c1b55d88e6816c9c8fb8f30565e04d37660cc73/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L100)

No need of removing them, these are brought about openmrsFormatter code style. i think we had talked about  code conventions in RATEST-74  for openmrs so when you again tell me to remove them, i really dont understand that. thanks

**Comment by kdaud** [comment]
> These are brought about openmrsFormatter code style.

Not sure why my  OpenMRS formatter does not create redundant code lines between all declared variables, as well between a method declaration and the first method variable!

Would love to learn from what @ibacher and @dkayiwa have to say about it !

**Comment by sherrif10** [comment]
Resolved thanks.

**Comment by kdaud** [comment]
> These are brought about openmrsFormatter code style.

Not sure why my  OpenMRS formatter does not create redundant code lines between all declared variables, as well between a method declaration and the first method variable!

Would love to learn from what @ibacher and @dkayiwa have to say about it !

**Comment by sherrif10** [comment]
Resolved thanks

**Comment by kdaud** [comment]
> These are brought about openmrsFormatter code style.

Not sure why my  OpenMRS formatter does not create redundant code lines between all declared variables as I see within FormTest file , as well between a method declaration and the first method variable!

Would love to learn from what @ibacher and @dkayiwa have to say about it !

**Comment by sherrif10** [comment]
Hello @kdaud  as you have seen these errors here https://github.com/openmrs/openmrs-distro-referenceapplication/pull/524/checks?check_run_id=3297720711  , its giving a confidence that there is something wrong with headerPage.Like i had said above we need to create a ticket that will handle headerPages.

**Comment by kdaud** [comment]
One thing one has to know with Github Actions is that some times fails to pick the online instance to run changes especially when the server is not stable. So as I see the logs, your changes are innocents about it and should not be condemned. Just trigger an empty commit to rebuild the changes and green will show. Take care @sherrif10

**Comment by sherrif10** [comment]
This is going to fail on bamboo as well with the exact errors logs. The recent commit was fine with github actions but failed on bamboo ,so ideally if its failing on githubs, absolutely it must fail on bamboo.

**Comment by kdaud** [comment]
If I was @sherrif10 I would try out the [recommendation](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/524#issuecomment-896554925)

**Comment by sherrif10** [comment]
> If I was @sherrif10 I would try out the [recommendation](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/524#issuecomment-896554925)

Thanks @kdaud  , As you can see here that  https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/HeaderPage.java#L22  , that we actually nolonger need that functionality in headerPage. Feel free to check its element to comfirm. Again this functionality is returning back to homePage where by HomePage.go(); does the same thing. Not sure if am clear.

**Comment by kdaud** [comment]
@sherrif10 you seem to have done something quit different in reference to the recommendation [here ](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/524#issuecomment-896554925). I do not think  we need to invoke the method you have introduced in the commit with id  `55d68a3` ie `headerPage.clickOnHomeLink();` is good enough and Bamboo will fail the test no more!

Its worth looking !!

**Comment by kdaud** [comment]
Have you just forgotten to omit redundant line [20](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/55d68a32b31fdce639f1f47e6df26560da727675/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L20) [65](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/55d68a32b31fdce639f1f47e6df26560da727675/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L65) [79](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/55d68a32b31fdce639f1f47e6df26560da727675/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L79) [93](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/55d68a32b31fdce639f1f47e6df26560da727675/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L93) ? You can resolve that and we welcome FormTest into master

**Comment by kdaud** [comment]
Was it an oversight to forget line [20](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/55d68a32b31fdce639f1f47e6df26560da727675/ui-tests/src/test/java/org/openmrs/reference/FormTest.java#L20)?

**Comment by sherrif10** [comment]
Resolved thanks

File: ui-tests/src/test/java/org/openmrs/reference/FormTest.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/FormTest.java:99]: We can as well here embrace a similar code indention which you have smartly addressed within `deletePatient()` method in the commit with id `79f836f`

**Comment by sherrif10** [comment]
Resolved thanks.

**Comment by kdaud** [comment]
Great effort @sherrif10 !

One last comment would be renaming the commit title from `Add Delete Functionality in Tear down method` to something that reflect what the commit has addressed.

**Comment by sherrif10** [comment]
Thanks @kdaud ,Resolved

**Comment by kdaud** [comment]
@sherrif10 any reason of making commit message Capitalized ? Its a bit shouting!

**Comment by sherrif10** [comment]
Actually i think Most of the title should be case sensitive to be more readable and probably pick java convention style

**Comment by kdaud** [comment]
> Actually i think Most of the title should be case sensitive to be more readable

You may need to take a survey on the existing `commit history` within your local master repo by running a simple command `git log`  then compare and contrast!

**Comment by kdaud** [comment]
You need to change the title manually as you did before not through a PR

**Comment by sherrif10** [comment]
The way i did should be the right way to do, and git provides that [resource](https://linuxize.com/post/change-git-commit-message/).

**Review by kdaud — APPROVED** [review]
Am keeping an eye to welcome FormTest into master after a merge that is awaiting for @dkayiwa  or @ibacher to perform!

**Pull request merged**
